File: backend/app/schemas/tra.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class TraResourceListResponse(BaseModel):
    items: list[dict[str, Any]]
    total_count: int


class TraBatchRowInput(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: int | None = None
    _status: str = "clean"


class TraResourceBatchRequest(BaseModel):
    items: list[TraBatchRowInput]


class TraResourceBatchResponse(BaseModel):
    created: int
    updated: int
    deleted: int


class TraGenerateRequiredEventsRequest(BaseModel):
    year: int = Field(ge=2000, le=2100)


class TraGenerateRequiredTargetsRequest(BaseModel):
    year: int = Field(ge=2000, le=2100)
    rule_code: str | None = Field(default=None, max_length=30)


class TraGenerateElearningWindowsRequest(BaseModel):
    year: int = Field(ge=2000, le=2100)
    app_count: int = Field(default=2, ge=0, le=31)


class TraApplyCyberResultsRequest(BaseModel):
    upload_ym: str | None = Field(default=None, pattern=r"^\d{6}$")


class TraGenerationResponse(BaseModel):
    processed: int
    message: str


# ─── Write-flow schemas ───────────────────────────────────────────────────────


class TraApplicationItem(BaseModel):
    id: int
    application_no: str
    employee_id: int
    employee_no: str | None
    employee_name: str | None
    department_name: str | None
    course_id: int
    course_name: str | None
    event_id: int | None
    event_name: str | None
    in_out_type: str | None
    status: str
    year_plan_yn: bool
    survey_yn: bool
    edu_memo: str | None
    note: str | None
    created_at: datetime
    updated_at: datetime


class TraApplicationListResponse(BaseModel):
    items: list[TraApplicationItem]
    total_count: int


class TraApplicationCreateRequest(BaseModel):
    course_id: int
    event_id: int | None = None
    in_out_type: str | None = None
    year_plan_yn: bool = False
    edu_memo: str | None = Field(default=None, max_length=2000)
    note: str | None = Field(default=None, max_length=2000)


class TraApplicationRejectRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=2000)


class TraApplicationActionResponse(BaseModel):
    item: TraApplicationItem

File: frontend/src/types/tra.ts
export type TraRowStatus = "clean" | "added" | "updated" | "deleted";

export type TraResourceRow = {
  id: number;
  _status: TraRowStatus;
  _original?: Record<string, unknown>;
  _prevStatus?: TraRowStatus;
  [key: string]: unknown;
};

export type TraResourceListResponse = {
  items: Record<string, unknown>[];
  total_count: number;
};

export type TraResourceBatchPayload = {
  items: Record<string, unknown>[];
};

export type TraResourceBatchResponse = {
  created: number;
  updated: number;
  deleted: number;
};

export type TraGenerationResponse = {
  processed: number;
  message: string;
};

// ── Write-flow types ──────────────────────────────────────────────────────────

export type TraApplicationItem = {
  id: number;
  application_no: string;
  employee_id: number;
  employee_no: string | null;
  employee_name: string | null;
  department_name: string | null;
  course_id: number;
  course_name: string | null;
  event_id: number | null;
  event_name: string | null;
  in_out_type: string | null;
  status: string;
  year_plan_yn: boolean;
  survey_yn: boolean;
  edu_memo: string | null;
  note: string | null;
  created_at: string;
  updated_at: string;
};

export type TraApplicationListResponse = {
  items: TraApplicationItem[];
  total_count: number;
};

export type TraApplicationCreateRequest = {
  course_id: number;
  event_id?: number | null;
  in_out_type?: string | null;
  year_plan_yn?: boolean;
  edu_memo?: string | null;
  note?: string | null;
};

export type TraApplicationRejectRequest = {
  reason?: string | null;
};

export type TraApplicationActionResponse = {
  item: TraApplicationItem;
};

export type TraColumnType = "text" | "number" | "boolean" | "date";

export type TraColumnConfig = {
  field: string;
  headerName: string;
  type?: TraColumnType;
  minWidth?: number;
  width?: number;
  flex?: number;
  editable?: boolean;
  required?: boolean;
};

export type TraSearchFieldConfig = {
  key: string;
  placeholder: string;
};

export type TraScreenConfig = {
  title: string;
  resource: string;
  searchFields: TraSearchFieldConfig[];
  columns: TraColumnConfig[];
  defaultRow: Record<string, unknown>;
};

// ─── Write-flow types ─────────────────────────────────────────────────────────

export type TraApplicationItem = {
  id: number;
  application_no: string;
  employee_id: number;
  employee_no: string | null;
  employee_name: string | null;
  department_name: string | null;
  course_id: number;
  course_name: string | null;
  event_id: number | null;
  event_name: string | null;
  in_out_type: string | null;
  status: string;
  year_plan_yn: boolean;
  survey_yn: boolean;
  edu_memo: string | null;
  note: string | null;
  created_at: string;
  updated_at: string;
};

export type TraApplicationListResponse = {
  items: TraApplicationItem[];
  total_count: number;
};

export type TraApplicationCreateRequest = {
  course_id: number;
  event_id?: number | null;
  in_out_type?: string | null;
  year_plan_yn?: boolean;
  edu_memo?: string | null;
  note?: string | null;
};

export type TraApplicationRejectRequest = {
  reason?: string | null;
};

export type TraApplicationActionResponse = {
  item: TraApplicationItem;
};
